src/pkg/cli/client/byoc/aws/subscribe.go

```diff
@@ -0,0 +1,249 @@
+		if state == defangv1.ServiceState_NOT_SPECIFIED {
+			return // Ignore unknown kaniko task states
+		}
+		if state == defangv1.ServiceState_BUILD_STOPPING {
+			for _, container := range evt.Containers {
+				if container.ExitCode != 0 {
+					state = defangv1.ServiceState_BUILD_FAILED
+				}
+			}
+		}
+		status = "BUILD_" + evt.LastStatus + " " + evt.StoppedReason
+	} else {
+		if etag != s.etag {
+			return
+		}
+
+		if evt.LastStatus == "DEACTIVATING" {
+			state = defangv1.ServiceState_DEPLOYMENT_FAILED // Fast deployment fail
+		} else {
+			state = defangv1.ServiceState_DEPLOYMENT_PENDING // Treat all other task updates as deployment pending
+		}
+
+		status = "TASK_" + evt.LastStatus
+		if evt.StoppedReason != "" {
+			status += " " + evt.StoppedReason
+		}
+	}
+
+	if !slices.Contains(s.services, service) {
+		return
+	}
+
+	if etag != "" && etag != s.etag {
+		return
+	}
+
+	s.send(&defangv1.SubscribeResponse{
+		Name:   service,
+		Status: status,
+		State:  state,
+	})
+}
+
+func (s *byocSubscribeServerStream) HandleEcsDeploymentStateChange(evt ecs.ECSDeploymentStateChange, resources []string) {
+	etag, ok := deploymentEtags[evt.DeploymentId]
+	if !ok {
+		return // Ignore deployments that are not associated with an etag
+	}
+	if etag != "" && etag != s.etag {
+		return // Ignore deployments that is not associated with the current etag
+	}
+
+	if len(resources) < 1 {
+		s.error(fmt.Errorf("error parsing ECS deployment state change: no resources"))
+		return
+	}
+
+	service, err := GetEcsDeploymetStateChangeService(resources)
+	if err != nil {
+		s.error(err)
+		return
+	}
+
+	var state defangv1.ServiceState
+	switch evt.EventName {
+	case "SERVICE_DEPLOYMENT_COMPLETED":
+		state = defangv1.ServiceState_DEPLOYMENT_COMPLETED
+	case "SERVICE_DEPLOYMENT_FAILED":
+		state = defangv1.ServiceState_DEPLOYMENT_FAILED
+	default:
+		state = defangv1.ServiceState_DEPLOYMENT_PENDING
+	}
+
+	status := evt.EventName
+	if evt.Reason != "" {
+		status += " " + evt.Reason
+	}
+
+	s.send(&defangv1.SubscribeResponse{
+		Name:   service,
+		Status: status,
+		State:  state,
+	})
+}
+
+func (s *byocSubscribeServerStream) Close() error {
+	s.closingLock.Lock()
+	defer s.closingLock.Unlock()
+	s.closed = true
+	close(s.ch)
+	return nil
+}
+
+func (s *byocSubscribeServerStream) Receive() bool {
+	select {
+	case resp, ok := <-s.ch:
+		if !ok || resp == nil {
+			return false
+		}
+		s.resp = resp
+		return true
+	case <-s.ctx.Done():
+		s.err = s.ctx.Err()
+		return false
+	}
+}
+
+func (s *byocSubscribeServerStream) Msg() *defangv1.SubscribeResponse {
+	return s.resp
+}
+
+func (s *byocSubscribeServerStream) Err() error {
+	return s.err
+}
+
+func (s *byocSubscribeServerStream) error(err error) {
+	s.err = err
+	s.send(nil)
+}
+
+func (s *byocSubscribeServerStream) send(resp *defangv1.SubscribeResponse) {
+	s.closingLock.Lock()
+	defer s.closingLock.Unlock()
+	if s.closed {
+		return
+	}
+	select {
+	case s.ch <- resp:
+	case <-s.ctx.Done():
```

Comment:
is the <-s.ctx.Done() necessary to detect an error occurring on the channel?

Reply:
This would prevent the goroutine waiting for the "Receive()" call be blocked when the context is cancelled.